src/pytorch/pme/pmeCUDA.cu

```diff
@@ -0,0 +1,447 @@
+    int maxExclusions = exclusions.size(1);
+    double energy = 0;
+
+    for (int index = blockIdx.x*blockDim.x+threadIdx.x; index < numNeighbors; index += blockDim.x*gridDim.x) {
+        int atom1 = neighbors[0][index];
+        int atom2 = neighbors[1][index];
+        float r = distances[index];
+        bool include = (atom1 > -1);
+        for (int j = 0; include && j < maxExclusions && exclusions[atom1][j] >= atom2; j++)
+            if (exclusions[atom1][j] == atom2)
+                include = false;
+        if (include) {
+            float invR = 1/r;
+            float alphaR = alpha*r;
+            float expAlphaRSqr = expf(-alphaR*alphaR);
+            float erfcAlphaR = erfcf(alphaR);
+            float prefactor = coulomb*invR;
+            float c1 = charge[atom1];
+            float c2 = charge[atom2];
+            energy += prefactor*erfcAlphaR*c1*c2;
+            atomicAdd(&chargeDeriv[atom1], prefactor*erfcAlphaR*c2);
+            atomicAdd(&chargeDeriv[atom2], prefactor*erfcAlphaR*c1);
+            float dEdR = prefactor*c1*c2*(erfcAlphaR+alphaR*expAlphaRSqr*M_2_SQRTPI)*invR*invR;
+            for (int j = 0; j < 3; j++) {
+                atomicAdd(&posDeriv[atom1][j], -dEdR*deltas[index][j]);
+                atomicAdd(&posDeriv[atom2][j], dEdR*deltas[index][j]);
+            }
+        }
+    }
+
+    // Subtract excluded interactions to compensate for the part that is
+    // incorrectly added in reciprocal space.
+
+    float dr[3];
+    for (int index = blockIdx.x*blockDim.x+threadIdx.x; index < numAtoms*maxExclusions; index += blockDim.x*gridDim.x) {
+        int atom1 = index/maxExclusions;
+        int atom2 = exclusions[atom1][index-atom1*maxExclusions];
+        if (atom2 > atom1) {
+            for (int j = 0; j < 3; j++)
+                dr[j] = pos[atom1][j]-pos[atom2][j];
+            float r2 = dr[0]*dr[0] + dr[1]*dr[1] + dr[2]*dr[2];
+            float invR = rsqrtf(r2);
+            float r = invR*r2;
+            float alphaR = alpha*r;
+            float expAlphaRSqr = expf(-alphaR*alphaR);
+            float erfAlphaR = erff(alphaR);
+            float prefactor = coulomb*invR;
+            float c1 = charge[atom1];
+            float c2 = charge[atom2];
+            energy -= prefactor*erfAlphaR*c1*c2;
+            atomicAdd(&chargeDeriv[atom1], -prefactor*erfAlphaR*c2);
+            atomicAdd(&chargeDeriv[atom2], -prefactor*erfAlphaR*c1);
+            float dEdR = prefactor*c1*c2*(erfAlphaR-alphaR*expAlphaRSqr*M_2_SQRTPI)*invR*invR;
+            for (int j = 0; j < 3; j++) {
+                atomicAdd(&posDeriv[atom1][j], dEdR*dr[j]);
+                atomicAdd(&posDeriv[atom2][j], -dEdR*dr[j]);
+            }
+        }
+    }
+    energyBuffer[blockIdx.x*blockDim.x+threadIdx.x] = energy;
+}
+
+__device__ void invertBoxVectors(const Accessor<float, 2>& box, float recipBoxVectors[3][3]) {
+    float determinant = box[0][0]*box[1][1]*box[2][2];
+    float scale = 1.0/determinant;
```

Comment:
1.0f so it does not get promoted to double and then converted back to float

Reply:
Done.